ERCS/erc-7786.md

```diff
@@ -0,0 +1,222 @@
+
+#### Post-processing
+
+After a sender has invoked `sendMessage`, further action MAY be required by the gateways to make the message effective. This is called *post-processing*. For example, some payment is typically required to cover the gas of executing the message at the destination.
+
+The exact interface for any such action is out of scope of this ERC. If the `postProcessingOwner` attribute is supported and present, such actions MUST be restricted to the specified account, otherwise they MUST be able to be performed by any party in a way that MUST NOT be able to compromise the eventual receipt of the message.
+
+The gateway MUST emit a `MessageSent` event with the appropriate identifier once no more post-processing is required by the source gateway. Further post-processing MAY be required in the source or destination gateways.
+
+### Destination Gateway
+
+A Destination Gateway is a contract that implements a protocol to validate messages sent on other chains and have them received at their destination.
+
+The gateway can operate in Active or Passive Mode.
+
+In both modes, the destination account of a message, aka the receiver, MUST implement a `receiveMessage` function. The use of this function depends on the mode of the gateway as described in the following sections.
+
+```solidity
+interface IERC7786Receiver {
+    function executeMessage(
+        address gateway,
+        bytes calldata gatewayMessageKey,
+        string calldata sourceChain, // CAIP-2 chain identifier
+        string calldata sender, // CAIP-10 account address
+        bytes calldata payload,
+        bytes[] calldata attributes
+    ) external payable;
+}
+```
+
+#### Active Mode
+
+The gateway directly invokes `receiveMessage`, and only does so with valid messages. The receiver MUST assume that a message is valid if the caller is a known gateway.
+
+The arguments `gateway` and `gatewayMessageKey` are unused in active mode and SHOULD be zero and empty respectively.
+
+#### Passive Mode
+
+The gateway does not directly invoke `receiveMessage`, but provides a means to validate messages. The receiver allows any party to invoke `receiveMessage`, but if the caller is not a known gateway it MUST check that the gateway provided as an argument is a known gateway, and it MUST validate the message against it before accepting it, forwarding the message key.
+
+A gateway acting in passive mode MUST implement `IERC7786GatewayDestinationPassive`. If a gateway operates exclusively in active mode, the implementation of this interface is OPTIONAL.
+
+```solidity
+interface IERC7786GatewayDestinationPassive {
+    function setMessageExecuted(
+        bytes calldata messageKey,
+        string calldata sourceChain, // CAIP-2 chain identifier
+        string calldata sender, // CAIP-10 account address
+        bytes calldata payload,
+        bytes[] calldata attributes
+    ) external;
+}
+```
+
+##### `setMessageExecuted`
```

Comment:
This is super confusing.

- It implies (but this is never stated explicitly) that messages have to be stored somewhere prior to this being called. Presumably by a previous call to `receiveMessage`?
      - This is super wasteful in terms of chain storage, and is expensive as well.
- I'm not entirely sure what the expected rationale is... a "gateway operator" wait for someone to call `receiveMessage` then actually executes it and calls `setMessageExecuted`?

Reply:
How are messages usually relayed/executed ?

There are cases such as mainnet->matic were the execution is part of the protocol, but in most cases, the execution involves some kind of ticket being passed and stored between chains.

For matic -> mainnet, we have a state root that is used to prove stuff that happened on matic. In that case the messageKey could be a merkle inclusion proof of some event in this state.

AFAIK, Arbitrum has a ticket system, which is basically the storage we need here. 

So yes its "wastefull", but unless the relayer has some specific authentication, it feels like its needed, and we already ahve this.

---

Now I get the possible confusion with the fact that the passive mode is optional, and only needs to be supported if the active mode has some failure mode. My personal opinion is that the passive mode can be hidden behind an "active adaptor". Many protocole already have a "retry mechanism" in the gateway. This "active adaptor" would turn a passive gateway into an active one with a similar "retry mechanism".

The reason we have this passive mode is that some protocole do work that way, and AFAIK they don't want to implement/maintain this active adaptor themselves. We could technically get rid of passive mode, and require someone to maintain that active adaptor ...

Reply:
Note: this passive mode is matching Axelar's workflow

"Someone/anyone" calls the destination contract with message details. This could use any endpoint, though there is a "standard" one that the axelar relayer networks uses, and that receiver that want use this network must have.

The receiver then calls axelar to confirm the details are correct. Axelar just says "yes or no". The receiver can fail, or do whatever they want.

Again the rational here is that the axelar gateway does never call any contract directly (for safety / reentrenacy protection)

Reply:
That probably doesn't need to be enshrined in the interface I think.

The ticketing system can be implementation-specific, a detail that only the gateway knows about. Then someone (whether the gateway in active mode or the user in passive mode) needs to call `receiveMessage` with the details (making it effectively similar to what `setMessageExecuted` currently does).

Reply:
> The ticketing system can be implementation-specific, a detail that only the gateway knows about. Then someone (whether the gateway in active mode or the user in passive mode) needs to call `receiveMessage` with the details

Unless I'm misunderstanding this is exactly what the current spec says.

`setMessageExecuted` (which was at some point called `validateReceivedMessage`) is needed for when the message is submitted via `receiveMessage` by an account that isn't the gateway, so the message needs to be validated with the gateway before the receiver accepts it.

Reply:
Added some diagrams.